Failure of postcondition shown as failure of the assertion in the branch

Trying to verify this:
```scala
import stainless.lang.* 
import stainless.collection.*
import stainless.annotation.*

def foldListConc[T,S](l1: List[T], l2: List[T], s0: S, seqop: (S,T) => S): Unit = {
  l1 match
    case Nil() => ()
    case Cons(t1,l1rest) => {
      assert(l1 ++ l2 == Cons(t1, l1rest ++ l2), "head of concat")
      assert((l1 ++ l2).foldLeft(s0)(seqop) == 
             (l1rest ++ l2).foldLeft(seqop(s0, t1))(seqop))
      foldListConc(l1rest, l2, s0, seqop)
    }
} ensuring(_ => (l1 ++ l2).foldLeft(s0)(seqop) == 
                l2.foldLeft(l1.foldLeft(s0)(seqop))(seqop))
```
gives the error that misleading suggests that an assert has failed, which is not the case, rather the postcondition fails:
```
warning:  - Result for 'postcondition' VC for foldListConc @11:7:
warning: l1.isInstanceOf[Nil] || ++[T](l1, l2) != Cons[T](l1.h, ++[T](l1.t, l2)) || foldLeft[T, S](++[T](l1, l2), s0, seqop) != foldLeft[T, S](++[T](l1.t, l2), seqop(s0, l1.h), seqop) || {
  val _$1: Unit = foldListConc[T, S](l1.t, l2, s0, seqop)
  foldLeft[T, S](++[T](l1, l2), s0, seqop) == foldLeft[T, S](l2, foldLeft[T, S](l1, s0, seqop), seqop)
}
PositionsError.scala:11:7: warning:  => TIMEOUT
                 assert(l1 ++ l2 == Cons(t1, l1rest ++ l2), "head of concat")
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

--- a/core/src/main/scala/stainless/ast/SymbolOps.scala
+++ b/core/src/main/scala/stainless/ast/SymbolOps.scala
@@ -172,7 +172,7 @@ trait SymbolOps extends inox.ast.SymbolOps with TypeOps { self =>
             case Some(g) => patCond withCond replaceFromSymbols(map, g)
             case None => patCond
           }
-          val newRhs = replaceFromSymbols(map, cse.rhs).copiedFrom(cse.rhs)
+          val newRhs = replaceFromSymbols(map, cse.rhs)
           (realCond.toClause.copiedFrom(cse), newRhs, cse)
         }
 

--- a/core/src/main/scala/stainless/verification/TypeChecker.scala
+++ b/core/src/main/scala/stainless/verification/TypeChecker.scala
@@ -648,8 +648,8 @@ class TypeChecker(val program: StainlessProgram, val context: inox.Context, val 
       case m: MatchExpr => inferType(tc, matchToIfThenElse(e, false))
 
       case IfExpr(b, e1, e2) =>
-        val (tpe1, tr1) = inferType(tc.withTruth(b).setPos(e1), e1)
-        val (tpe2, tr2) = inferType(tc.withTruth(Not(b)).setPos(e2), e2)
+        val (tpe1, tr1) = inferType(tc.withTruth(b).withLastPos(e1), e1)
+        val (tpe2, tr2) = inferType(tc.withTruth(Not(b)).withLastPos(e2), e2)
         (ite(b, tpe1, tpe2), checkType(tc.setPos(b).withVCKind(VCKind.CheckType), b, BooleanType()) ++ tr1 ++ tr2)
 
       case Error(tpe, descr) =>
@@ -717,7 +717,7 @@ class TypeChecker(val program: StainlessProgram, val context: inox.Context, val 
         val trValue = checkType(tc.setPos(value).withVCKind(VCKind.CheckType), value, vd.tpe)
         val (tc2, id2) = tc.freshBindWithValue(vd, value)
         val freshBody: Expr = Substituter(immutable.Map(vd.id -> id2)).transform(body)
-        val (tpe, trBody) = inferType(tc2.setPos(body), freshBody)
+        val (tpe, trBody) = inferType(tc2.withLastPos(body), freshBody)
         (insertFreshLets(Seq(vd), Seq(value), tpe), trValue ++ trBody)
 
       case Assume(cond, body) =>
@@ -1082,7 +1082,7 @@ class TypeChecker(val program: StainlessProgram, val context: inox.Context, val 
         val (tc2, id2) = tc.freshBindWithValue(vd, value)
         val freshBody: Expr = Substituter(immutable.Map(vd.id -> id2)).transform(body)
         checkType(tc.setPos(value).withVCKind(VCKind.CheckType), value, vd.tpe) ++
-        checkType(tc2.setPos(body), freshBody, tpe)
+        checkType(tc2.withLastPos(body), freshBody, tpe)
 
       case (Assert(cond, optErr, body), _) =>
         val kind = VCKind.fromErr(optErr)
@@ -1094,8 +1094,8 @@ class TypeChecker(val program: StainlessProgram, val context: inox.Context, val 
 
       case (IfExpr(b, e1, e2), _) =>
         checkType(tc.setPos(b).withVCKind(VCKind.CheckType), b, BooleanType()) ++
-        checkType(tc.withTruth(b).setPos(e1), e1, tpe) ++
-        checkType(tc.withTruth(Not(b)).setPos(e2), e2, tpe)
+        checkType(tc.withTruth(b).withLastPos(e1), e1, tpe) ++
+        checkType(tc.withTruth(Not(b)).withLastPos(e2), e2, tpe)
 
       case (e, TrueBoolean()) =>
         checkType(tc.withVCKind(VCKind.CheckType), e, BooleanType()) ++ buildVC(tc, e)

--- a/core/src/main/scala/stainless/verification/TypeCheckerContext.scala
+++ b/core/src/main/scala/stainless/verification/TypeCheckerContext.scala
@@ -101,6 +101,8 @@ object TypeCheckerContext {
       )
     }
 
+    def withLastPos(e: Expr): TypingContext = setPos(lastPos(e))
+
     def withTypeVariables(vars: Set[TypeParameter])(using opts: PrinterOptions, ctx: inox.Context): TypingContext = {
       checkFreshTypeVariables(vars)
       copy(typeVariables = typeVariables ++ vars).setPos(this)

--- a/core/src/main/scala/stainless/verification/TypeCheckerUtils.scala
+++ b/core/src/main/scala/stainless/verification/TypeCheckerUtils.scala
@@ -222,4 +222,14 @@ object TypeCheckerUtils {
   }
 
   def pred(e: Expr) = Minus(e, IntegerLiteral(1))
+
+  def lastPos(e: Expr): inox.utils.Position = e match {
+    case Assert(_, _, body) => lastPos(body)
+    case Assume(_, body) => lastPos(body)
+    case Require(_, body) => lastPos(body)
+    case Decreases(_, body) => lastPos(body)
+    case Ensuring(body, _) => lastPos(body)
+    case Let(_, _, body) => lastPos(body)
+    case _ => e.getPos
+  }
 }

